<!DOCTYPE html>
<html >
<head>

	<meta http-equiv="cache-control" content="no-cache, no-store, must-revalidate" />
	<meta http-equiv="expires" content="0" />
	<meta http-equiv="pragma" content="no-cache" />

	<!-- Due to a bug in webcomponents, I can't use CSS with HTML imports -->
	<!-- (at least in IE) -->
	<!-- https://github.com/webcomponents/webcomponentsjs/issues/400 -->
	<link rel="stylesheet" href="../../bootstrap/css/bootstrap-yeti.css"/>
	<link rel="stylesheet" href="../css/style.css"/>
	<link rel="stylesheet" href="./statsToolApp.css"/>


	<!-- Polyfill, required for browsers that do not support link rel="import" -->
	<script type="text/javascript" src="../../webcomponents/webcomponents-lite.min.js"></script>

	<link rel="import" href="./statsToolDependencies.html"/>
</head>
<body>


<div class="container" ng-app="statsToolApp">

	<div ng-controller="statsToolCtrl">
	
	
			<h2>Metagenomic Dark Contigs</h2>
		<div class="row">
			<div class="col-md-8"
				<p>
				When a metagenomics sample is sequenced and data is assembled, large numbers of contigs are generated. 
				Some of these are <b>known contigs</b> which match sequences already characterised in nucleotide or protein databases.
				Others are <b>dark contigs</b>: sequence fragments which are not homologous to any known sequence but have 
				simple characteristics indicating they may be of biological origin. It is even possible that some dark contigs may be 
				homologous across different metagenomic samples. Further investigation of these dark contigs may reveal new sequencing results.
				</p>
				<p>
				This simple online tool allows the exploration of a database containing both known and dark contigs from several metagenomics samples.
				</p>
			</div>
			<div class="col-md-4"
				<p><img alt="MRC-University of Glasgow Centre for Virus Research" src="images/CVRLogo.png"></p>
				<p><a href="http://www.gla.ac.uk/researchinstitutes/iii/cvr/" target="_blank"><small>MRC-University of Glasgow Centre for Virus Research</small></a></p>
			</div>
		</div>
		<p>
		Sample:
		<select ng-model="currentSample"
				ng-options="sample.Source for sample in samples">
		</select>
		Sequencing run:
		<select ng-disable="!sequences"
				ng-model="currentSequence"
				ng-options="sequence.ID for sequence in sequences">
		</select>
		<button type="button" class="btn btn-primary btn-xs"
			ng-click="showSequencingRunDetails()"
			ng-disabled="false">Sequencing run details</button>
		<button type="button" class="btn btn-primary btn-xs"
			ng-click="getContigs()"
			ng-disabled="false">Retrieve contig data</button>
		</p>
		<div ng-show="contigs.length > 0">
			<p>
			X metric:
			<select ng-model="xMetric"
					ng-options="metric.description for metric in sequenceMetrics">
			</select>
			Y metric:
			<select ng-model="yMetric"
					ng-options="metric.description for metric in sequenceMetrics">
			</select>
			</p>
			
			
			<div class="row">
			    <div class="col-md-6">
						<svg ng-attr-width="{{svgWidth}}" ng-attr-height="{{svgHeight}}"></svg>
			    </div>
			    <div ng-show="contigDetails" class="col-md-6">
			    	<h3>{{contigDetails.title}}</h3>
			    	<p><small>{{contigDetails.contigId}}</small></p>
					<table class="table table-striped table-bordered"
						style="table-layout: fixed; word-wrap: break-word;">
						<thead>
							<col width="28%"></col>
							<col width="72%"></col>
						</thead>
						<tr ng-repeat="prop in contigDetails.properties track by $index">
							<td><b>{{prop.description}}</b></td>
							<td>
								<div ng-switch="prop.linkType">
									<div ng-switch-when="href">
										<a href="{{prop.linkValue[0]}}" target="_blank"}}>{{prop.value}}</a>
									</div>
									<div ng-switch-when="relatedDark">
										<a ng-click="showRelatedDark(prop.linkValue[0], prop.linkValue[1])">{{prop.value}}</a>
									</div>
									<div ng-switch-default>
										{{prop.value}}
									</div>
								</div>
							</td>
						</tr>
					</table>
			    </div>
			</div>
			
			
			
			
		</div>
	</div>

</div>

</body>
</html>